python/metatensor-operations/metatensor/operations/to.py

```diff
@@ -116,19 +120,81 @@ def block_to(
+    # Walk the tree of gradients without recursion
+    # (recursion is not supported by torchscript)
+    current_location: List[str] = []
+    current_block = block
+    transformed_blocks: List[TensorBlock] = []
+    last_visited = ""
+
+    while True:
+        print(len(transformed_blocks))
+        gradient_names = current_block.gradients_list()
+        n_gradients = len(gradient_names)
+        if last_visited == "":  # first time we see this block
+            # transform and append to list of transformed blocks:
+            transformed_blocks.append(
+                _block_to(current_block, backend, dtype, device, requires_grad)
+            )
+            if n_gradients == 0:  # the current block has no gradients
+                # step back:
+                if len(current_location) == 0:
+                    break
+                last_visited = (
+                    current_location.pop()
+                )  # removes last visited gradient name and stores it
+                current_block = _reach_current_block(
+                    block, current_location
+                )  # reach current location
+            else:
+                # proceed walking forward:
+                current_block = current_block.gradient(gradient_names[0])
+                current_location.append(gradient_names[0])
+        else:  # we're coming back to a block we've already seen
+            index_last_visited = gradient_names.index(last_visited)
+            if (
+                index_last_visited == n_gradients - 1
+            ):  # the last visited gradient was the last one we needed to convert
+                # add gradients to the current gradient:
+                for i_gradient in range(n_gradients):
+                    transformed_blocks[-n_gradients - 1].add_gradient(
+                        gradient_names[i_gradient],
+                        transformed_blocks[i_gradient - n_gradients],
+                    )
+                for _ in range(n_gradients):
+                    # remove all added gradients from the transformed list:
+                    transformed_blocks.pop()
+                # step back:
+                if len(current_location) == 0:
+                    break
+                last_visited = (
+                    current_location.pop()
+                )  # removes last visited gradient and stores it
+                current_block = _reach_current_block(
+                    block, current_location
+                )  # reach current location
+            else:  # more gradients to convert
+                # walk forward:
+                current_block = current_block.gradient(
+                    gradient_names[index_last_visited + 1]
+                )
+                current_location.append(gradient_names[index_last_visited + 1])
+                last_visited = ""
+
+    # at this point, transformed_blocks should only contain the final transformed block:
+    return transformed_blocks[0]
```

Comment:
what about the other elements of `transformed_blocks`?

Reply:
They will all have been popped from the list and added to the first one as gradients